### PR TITLE
Towards a single code-path for enumeration + dev_open

### PR DESCRIPTION
**Pull request by safl**
This PR is an integration of the following:

* https://github.com/OpenMPDK/xNVMe/pull/125
* https://github.com/OpenMPDK/xNVMe/pull/120
  - Since #126 does a complete rewrite of the controller-reference-logic, then this specific change does not apply anymore
* https://github.com/OpenMPDK/xNVMe/pull/114
* https://github.com/OpenMPDK/xNVMe/pull/110
  - missing the CIJOE tests from this PR
* https://github.com/OpenMPDK/xNVMe/pull/102

This, seem to solve the multi-subsystem on fabrics-endpoint issues:

```
safl@debtop:~/git/xnvme$ sudo xnvme enum --uri 106.110.32.96:4420
xnvme_enumeration:
  - {uri: '106.110.32.96:4420', dtype: 0x2, nsid: 0x1, csi: 0x0, subnqn: 'nqn.2022-06.io.xnvme:ctrlnode1'}
  - {uri: '106.110.32.96:4420', dtype: 0x2, nsid: 0x1, csi: 0x0, subnqn: 'nqn.2022-06.io.xnvme:ctrlnode3'}
  - {uri: '106.110.32.96:4420', dtype: 0x2, nsid: 0x1, csi: 0x0, subnqn: 'nqn.2022-06.io.xnvme:ctrlnode2'}
  - {uri: '106.110.32.96:4420', dtype: 0x2, nsid: 0x1, csi: 0x0, subnqn: 'nqn.2022-06.io.xnvme:ctrlnode4'}
safl@debtop:~/git/xnvme$ sudo xnvme info 106.110.32.96:4420 --dev-nsid 0x1 --subnqn "nqn.2022-06.io.xnvme:ctrlnode3"
xnvme_dev:
  xnvme_ident:
    uri: '106.110.32.96:4420'
    dtype: 0x2
    nsid: 0x1
    csi: 0x0
    subnqn: 'nqn.2022-06.io.xnvme:ctrlnode3'
  xnvme_be:
    admin: {id: 'nvme'}
    sync: {id: 'nvme'}
    async: {id: 'nvme'}
    attr: {name: 'spdk'}
  xnvme_opts:
    be: 'spdk'
    mem: 'FIX-ID-VS-MIXIN-NAME'
    dev: 'FIX-ID-VS-MIXIN-NAME'
    admin: 'nvme'
    sync: 'nvme'
    async: 'nvme'
    oflags: 0x4
  xnvme_geo:
    type: XNVME_GEO_CONVENTIONAL
    npugrp: 1
    npunit: 1
    nzone: 1
    nsect: 28131328
    nbytes: 512
    nbytes_oob: 0
    tbytes: 14403239936
    mdts_nbytes: 131072
    lba_nbytes: 512
    lba_extended: 0
    ssw: 9
```

Things to address before merging this:

- [x] fix regressions
- [x] docs: the text needs an update; it is currently describing the old scripts
- [x] Fix the ``fabrics_target_spdk.sh`` it is currently not constructing the subsystem correctly, it is listening but when using it, either via ``nvme-cli`` or ``xnvme`` then the second device is not showing up.
- [x] Update commit-messages as they are not describing what they do
- [ ] Integrate tests from https://github.com/OpenMPDK/xNVMe/pull/110

### CODE REVIEW

**Comment by safl** [comment]
Hi @LiadOz,
Can you try out this PR? The controller-logic is rewritten and your approach to "refreshing" the controller-instance applied.
Additionally, ``subnqn`` is added where needed.
Best,
Simon

**Comment by LiadOz** [comment]
@safl I think that xnvme_dev_open does not work correctly, I'm trying to run `xnvme_dev 127.0.0.1:4420` with local spdk server setup but it doesn't find the device, running  xnvme enum works.
```
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'spdk'
# DBG:xnvme_be.c:xnvme_be_factory-674: INFO: obtained backend instance be: 'spdk'
# DBG:xnvme_be_spdk_dev.c:_spdk_env_init-185: INFO: SPDK NVMe PCIe Driver registration -- BEGIN
# DBG:xnvme_be_spdk_dev.c:_spdk_env_init-190: INFO: skipping, already registered.
# DBG:xnvme_be_spdk_dev.c:_spdk_env_init-192: INFO: SPDK NVMe PCIe Driver registration -- END
[2022-07-14 15:06:58.862164] Starting SPDK v21.10 git sha1 4e4f11ff7 / DPDK 21.08.0 initialization...
[2022-07-14 15:06:58.862354] [ DPDK EAL parameters: [2022-07-14 15:06:58.862435] xnvme [2022-07-14 15:06:58.862541] -c 0x1 [2022-07-14 15:06:58.862619] --log-level=lib.eal:6 [2022-07-14 15:06:58.862690] --log-level=lib.cryptodev:5 [2022-07-14 15:06:58.862760] --log-level=user1:6 [2022-07-14 15:06:58.862830] --iova-mode=pa [2022-07-14 15:06:58.862875] --base-virtaddr=0x200000000000 [2022-07-14 15:06:58.862908] --match-allocations [2022-07-14 15:06:58.862938] --file-prefix=spdk0 [2022-07-14 15:06:58.862968] --proc-type=auto [2022-07-14 15:06:58.863002] ]
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
# DBG:xnvme_be_spdk_dev.c:_spdk_env_init-218: INFO: spdk_env_is_initialized: 1
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-575: INFO: probing START
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-584: INFO: do probe of trtype: PCIe, # 1 of 4 ?
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-622: INFO: yes; constructed trid, probing ...
EAL: Cannot find device (0127:00:00.1)
EAL: Failed to attach device on primary process
[2022-07-14 15:06:59.006304] nvme.c: 838:nvme_probe_internal: *ERROR*: NVMe ctrlr scan failed
[2022-07-14 15:06:59.006327] nvme.c: 915:spdk_nvme_probe: *ERROR*: Create probe context failed
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-625: FAILED: spdk_nvme_probe(), err: -1
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-584: INFO: do probe of trtype: TCP, # 2 of 4 ?
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-622: INFO: yes; constructed trid, probing ...
[2022-07-14 15:06:59.582536] nvme_ctrlr.c: 704:nvme_ctrlr_set_intel_support_log_pages: *WARNING*: [nqn.2016-06.io.spdk:cnode1] Intel log pages not supported on Intel drive!
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-584: INFO: do probe of trtype: RDMA, # 3 of 4 ?
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-622: INFO: yes; constructed trid, probing ...
[2022-07-14 15:06:59.582650] nvme.c: 830:nvme_probe_internal: *ERROR*: NVMe trtype 1 not available
[2022-07-14 15:06:59.582665] nvme.c: 915:spdk_nvme_probe: *ERROR*: Create probe context failed
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-625: FAILED: spdk_nvme_probe(), err: -1
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-584: INFO: do probe of trtype: FC, # 4 of 4 ?
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-592: INFO: no; skipping due to ENOSYS
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-628: INFO: probing DONE
# DBG:xnvme_be_spdk_dev.c:xnvme_be_spdk_dev_open-685: INFO: no-match for TCP|RDMA
# DBG:xnvme_be.c:xnvme_be_factory-694: INFO: skipping backend due to err: -22
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'linux'
# DBG:xnvme_be.c:xnvme_be_factory-674: INFO: obtained backend instance be: 'linux'
# DBG:xnvme_be_linux_dev.c:xnvme_be_linux_dev_open-51: INFO: open() : opts->oflags: 0x4, flags: 0x2, opts->create_mode: 0x180
# DBG:xnvme_be_linux_dev.c:xnvme_be_linux_dev_open-56: FAILED: open(uri: '127.0.0.1:4420') : state->fd: '-1', errno: 2
# DBG:xnvme_be_linux_dev.c:xnvme_be_linux_dev_open-63: FAILED: open(uri: '127.0.0.1:4420') : state->fd: '-1', errno: 2
# DBG:xnvme_be.c:xnvme_be_factory-694: INFO: skipping backend due to err: -2
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'fbsd'
# DBG:xnvme_be.c:xnvme_be_factory-646: INFO: skipping be: 'fbsd'; !enabled
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'posix'
# DBG:xnvme_be.c:xnvme_be_factory-674: INFO: obtained backend instance be: 'posix'
# DBG:xnvme_be_posix_dev.c:xnvme_be_posix_dev_open-46: INFO: open() : opts->oflags: 0x4, flags: 0x2, opts->create_mode: 0x180
# DBG:xnvme_be_posix_dev.c:xnvme_be_posix_dev_open-51: FAILED: open(uri: '127.0.0.1:4420'), state->fd: '-1', errno: 2
# DBG:xnvme_be.c:xnvme_be_factory-694: INFO: skipping backend due to err: -2
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'macos'
# DBG:xnvme_be.c:xnvme_be_factory-646: INFO: skipping be: 'macos'; !enabled
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'windows'
# DBG:xnvme_be.c:xnvme_be_factory-646: INFO: skipping be: 'windows'; !enabled
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'ramdisk'
# DBG:xnvme_be.c:xnvme_be_factory-674: INFO: obtained backend instance be: 'ramdisk'
# DBG:xnvme_be_ramdisk_dev.c:_xnvme_be_ramdisk_dev_get_size-45: FAILED: Invalid URI. Only postfix of 'GB' allowed: 127.0.0.1:4420
# DBG:xnvme_be.c:xnvme_be_factory-694: INFO: skipping backend due to err: -22
# DBG:xnvme_be.c:xnvme_be_factory-643: INFO: checking be: 'vfio'
# DBG:xnvme_be.c:xnvme_be_factory-674: INFO: obtained backend instance be: 'vfio'
nvme_init (subprojects/libvfn/src/nvme/core.c:443): could not get device class code
# DBG:xnvme_be_vfio_dev.c:xnvme_be_vfio_dev_open-192: FAILED: initializing nvme device: 2
# DBG:xnvme_be.c:xnvme_be_factory-694: INFO: skipping backend due to err: -2
# DBG:xnvme_be.c:xnvme_be_factory-699: FAILED: no backend for uri: '127.0.0.1:4420'
# DBG:xnvme_dev.c:xnvme_dev_open-158: FAILED: failed opening uri: 127.0.0.1:4420
xnvme_dev: ~
```

**Comment by safl** [comment]
@LiadOz you have to provide the --subnqn. Via api: opts.subnqn, via cli: "xnvme info --dev-nsid 1 --subnqn ..."

**Comment by LiadOz** [comment]
I am having regression issues with this branch, but I am not able to find a small test to reproduce and it fails sporadically. Do you know if there are changes that might cause this to happen?
I also noticed that there is no removal of closed controllers and refcount isn't really used.

**Review by karlowich — COMMENTED** [review]
I have left a comment regarding the probing.
Another thing, since the probing code in `xnvme_be_spdk_dev_open` and `xnvme_be_spdk_enumerate` is identical maybe we should do separate function for that part?

File: lib/xnvme_be_spdk_dev.c [hunk 1]
```diff
@@ -233,15 +194,18 @@ _spdk_env_init(struct spdk_env_opts *opts)
 
 	xnvme_spdk_choker_on();
 
-	if (!opts) {
-		opts = &_spdk_opts;
+	spdk_env_opts_init(&env_opts);
+	env_opts.name = "xnvme";
+	env_opts.shm_id = 0;
```
Comment (LiadOz) [inline lib/xnvme_be_spdk_dev.c:199]: This line should be removed, spdk_env_opts_init sets this to -1 which will configure dpdk with the flag --no-shconf. using shm_id=0 will setup multiprocess environment, you may need to move it to inside the if block

**Comment by safl** [comment]
Closing this without merge since the individual changes have gone in without the refactoring.
It would still be good to improve, so adding to backlog.

**Pull request closed without merging**